haystack/nodes/retriever/web.py

```diff
@@ -319,18 +204,72 @@ def scrape_link_content(link: SearchResult) -> List[Document]:
 
             return docs
 
-        thread_count = cpu_count() if len(links) > cpu_count() else len(links)
+        thread_count = min(cpu_count() if len(links) > cpu_count() else len(links), 10)  # max 10 threads
         with ThreadPoolExecutor(max_workers=thread_count) as executor:
-            scraped_pages: Iterator[List[Document]] = executor.map(scrape_link_content, links)
+            fetched_pages: Iterator[List[Document]] = executor.map(link_fetch, links)
 
         # Flatten list of lists to a single list
-        extracted_docs = [doc for doc_list in scraped_pages for doc in doc_list]
+        extracted_docs = [doc for doc_list in fetched_pages for doc in doc_list]
 
         # Sort by score
         extracted_docs = sorted(extracted_docs, key=lambda x: x.meta["search.score"], reverse=True)
 
         return extracted_docs
 
+    def _check_cache(self, links: List[SearchResult]) -> Tuple[List[SearchResult], List[Document]]:
+        """
+        Check the DocumentStore cache for documents.
+
+        :param links: List of SearchResult objects.
+        :return: Tuple of lists of SearchResult and Document objects that were found in the cache.
+        """
+        if not links or not self.cache_document_store:
+            return [], []
+
+        cache_documents: List[Document] = []
+        cached_links: List[SearchResult] = []
+
+        cache_ids = [link.url for link in links if link.url]
+        for idx, cache_id in enumerate(cache_ids):
+            cache_filter: FilterType = {"url": cache_id}
+            documents = self.cache_document_store.get_all_documents(filters=cache_filter, return_embedding=False)
+            if documents:
+                cache_documents.extend(documents)
+                cached_links.append(links[idx])
```

Comment:
The current implementation is wrong. `cache_ids` is a filtered list of `links`, we risk caching the wrong stuff and even out of bounds exceptions.

This should solve the issue.

```suggestion
        valid_links = [link for link in links if link.url]
        for link in valid_links:
            cache_filter: FilterType = {"url": link.url}
            documents = self.cache_document_store.get_all_documents(filters=cache_filter, return_embedding=False)
            if documents:
                cache_documents.extend(documents)
                cached_links.append(link)
```

Reply:
Oh crap, you are right 🙏